plugin.program.AML/resources/main.py

```diff
@@ -0,0 +1,6751 @@
+        catalog_dic = fs_get_cataloged_dic_all(PATHS, catalog_name)
+    elif mame_view_mode == VIEW_MODE_PCLONE:
+        catalog_dic = fs_get_cataloged_dic_parents(PATHS, catalog_name)
+    if not catalog_dic:
+        kodi_dialog_OK('Catalog is empty. Rebuild the MAME databases.')
+        xbmcplugin.endOfDirectory(handle = g_addon_handle, succeeded = True, cacheToDisc = False)
+        return
+
+    loading_ticks_end = time.time()
+    rendering_ticks_start = time.time()
+    for catalog_key in sorted(catalog_dic):
+        if mame_view_mode == VIEW_MODE_FLAT:
+            num_machines = cache_index_dic[catalog_name][catalog_key]['num_machines']
+            if num_machines == 1: machine_str = 'machine'
+            else:                 machine_str = 'machines'
+        elif mame_view_mode == VIEW_MODE_PCLONE:
+            num_machines = cache_index_dic[catalog_name][catalog_key]['num_parents']
+            if num_machines == 1: machine_str = 'parent'
+            else:                 machine_str = 'parents'
+        _render_catalog_list_row(catalog_name, catalog_key, num_machines, machine_str)
+    xbmcplugin.endOfDirectory(handle = g_addon_handle, succeeded = True, cacheToDisc = False)
+    rendering_ticks_end = time.time()
+
+    # --- DEBUG Data loading/rendering statistics ---
+    log_debug('Loading seconds   {0}'.format(loading_ticks_end - loading_ticks_start))
+    log_debug('Rendering seconds {0}'.format(rendering_ticks_end - rendering_ticks_start))
+
+def _render_catalog_list_row(catalog_name, catalog_key, num_machines, machine_str):
+    # --- Create listitem row ---
+    ICON_OVERLAY = 6
+    title_str = '{0} [COLOR orange]({1} {2})[/COLOR]'.format(catalog_key, num_machines, machine_str)
+    plot_str = 'Catalog {0}\nCategory {1}'.format(catalog_name, catalog_key)
+    listitem = xbmcgui.ListItem(title_str)
+    listitem.setInfo('video', {'title' : title_str,      'plot' : plot_str,
+                               'overlay' : ICON_OVERLAY, 'size' : num_machines})
+
+    # --- Artwork ---
+    icon_path   = AML_ICON_FILE_PATH.getPath()
+    fanart_path = AML_FANART_FILE_PATH.getPath()
+    listitem.setArt({'icon' : icon_path, 'fanart' : fanart_path})
+
+    # --- Create context menu ---
+    URL_utils = _misc_url_3_arg_RunPlugin('command', 'UTILITIES',
+                                          'catalog', catalog_name, 'category', catalog_key)
+    commands = [
+        ('View', _misc_url_1_arg_RunPlugin('command', 'VIEW')),
+        ('Utilities', URL_utils),
+        ('Kodi File Manager', 'ActivateWindow(filemanager)'),
+        ('AML addon settings', 'Addon.OpenSettings({0})'.format(__addon_id__)),
+    ]
+    listitem.addContextMenuItems(commands)
+
+    # --- Add row ---
+    URL = _misc_url_2_arg('catalog', catalog_name, 'category', catalog_key)
+    xbmcplugin.addDirectoryItem(handle = g_addon_handle, url = URL, listitem = listitem, isFolder = True)
+
+#
+# Renders a list of parent MAME ma
```

Comment:
use a builtin instead of xbmc.Player()

Reply:
I don't understand what you mean. Can you please elaborate your comment again?

Reply:
Ups sorry for the lack of info. The Player class should not be used by plugins but only in scripts (where you create a custom interface). This is mainly due to the `play()` method which will save every file played/url to the database. Plugins should use setResolvedUrl instead so that a `plugin://` path is saved into the database instead. In this case you don't use the Player class to request media playback, however since this addon is a plugin I'd like to enforce some consistency in plugins. There's absolutely nothing wrong with your code it's basically a way to prevent others to state "but plugin x is using the Player" in the future.
Please replace `xbmc.Player().isPlaying` by `xbmc.getCondVisibility("Player.HasMedia")` if you want to detect the player is active or `xbmc.getCondVisibility("Player.Playing")` if you want to detect the player is actually playing (not paused, rewinding, etc).
`xbmc.Player().stop()` can be changed in favour of `xbmc.executebuiltin("PlayerControl(stop)")`. Pause to `xbmc.executebuiltin("PlayerControl(pause)")`. In alternative you can also use the json-rpc API: `xbmc.executeJSONRpc("{"jsonrpc":"2.0","method":"Player.Stop","params":{"playerid":0},"id":1}")

Reply:
Done. I have removed all instances of `xbmc.Player()` and replaced them with either `xbmc.getCondVisibility()` or `xbmc.executebuiltin()`.